File: process_osm.py
# DEPRECATED: Use process_osm_v2.py instead
import json
import sqlite3
import uuid
import random

with open('osm_data.json', 'r', encoding='utf-8') as f:
    data = json.load(f)

conn = sqlite3.connect('database.db')
cursor = conn.cursor()

# Clear existing EV data
cursor.execute("DELETE FROM ev_chargers")
cursor.execute("DELETE FROM stations WHERE type = 'ev'")

# District mapping from populate_portugal.py
PORTUGAL_MAPPING = {
    'Lisboa': ['Lisboa', 'Sintra', 'Cascais', 'Loures', 'Amadora', 'Oeiras', 'Odivelas', 'Vila Franca de Xira', 'Mafra', 'Torres Vedras', 'Alenquer', 'Lourinhã', 'Cadaval', 'Sobral de Monte Agraço', 'Arruda dos Vinhos'],
    'Porto': ['Porto', 'Vila Nova de Gaia', 'Matosinhos', 'Maia', 'Gondomar', 'Valongo', 'Póvoa de Varzim', 'Vila do Conde', 'Santo Tirso', 'Penafiel', 'Amarante', 'Felgueiras', 'Lousada', 'Paços de Ferreira', 'Paredes', 'Baião', 'Marco de Canaveses', 'Trofa'],
    'Setúbal': ['Setúbal', 'Almada', 'Seixal', 'Barreiro', 'Palmela', 'Moita', 'Sesimbra', 'Montijo', 'Alcochete', 'Grândola', 'Santiago do Cacém', 'Sines', 'Alcácer do Sal'],
    'Braga': ['Braga', 'Guimarães', 'Vila Nova de Famalicão', 'Barcelos', 'Fafe', 'Esposende', 'Vizela', 'Amares', 'Barcelos', 'Cabeceiras de Basto', 'Celorico de Basto', 'Esposende', 'Fafe', 'Guimarães', 'Póvoa de Lanhoso', 'Terras de Bouro', 'Vieira do Minho', 'Vila Nova de Famalicão', 'Vila Verde', 'Vizela'],
    'Aveiro': ['Aveiro', 'Santa Maria da Feira', 'Oliveira de Azeméis', 'Ovar', 'Ílhavo', 'Águeda', 'Estarreja', 'Anadia', 'Albergaria-a-Velha', 'Castelo de Paiva', 'Espinho', 'Murtosa', 'Oliveira do Bairro', 'Sever do Vouga', 'Vagos', 'Vale de Cambra'],
    'Leiria': ['Leiria', 'Alcobaça', 'Caldas da Rainha', 'Pombal', 'Marinha Grande', 'Peniche', 'Nazaré', 'Batalha', 'Bombarral', 'Castanheira de Pera', 'Figueiró dos Vinhos', 'Óbidos', 'Pedrógão Grande', 'Porto de Mós'],
    'Santarém': ['Santarém', 'Ourém', 'Tomar', 'Abrantes', 'Torres Novas', 'Entroncamento', 'Cartaxo', 'Almeirim', 'Alpiarça', 'Benavente', 'Chamusca', 'Constância', 'Coruche', 'Ferreira do Zêzere', 'Golegã', 'Mação', 'Rio Maior', 'Salvaterra de Magos', 'Sardoal', 'Vila Nova da Barquinha'],
    'Coimbra': ['Coimbra', 'Figueira da Foz', 'Cantanhede', 'Oliveira do Hospital', 'Condeixa-a-Nova', 'Arganil', 'Góis', 'Lousã', 'Mira', 'Miranda do Corvo', 'Montemor-o-Velho', 'Pampilhosa da Serra', 'Penacova', 'Penela', 'Soure', 'Tábua', 'Vila Nova de Poiares'],
    'Faro': ['Faro', 'Portimão', 'Loulé', 'Albufeira', 'Silves', 'Lagos', 'Tavira', 'Olhão', 'Vila Real de Santo António', 'Aljezur', 'Alcoutim', 'Castro Marim', 'Monchique', 'São Brás de Alportel', 'Vila do Bispo'],
    'Viseu': ['Viseu', 'Lamego', 'Mangualde', 'São Pedro do Sul', 'Tondela', 'Armamar', 'Carregal do Sal', 'Castro Daire', 'Cinfães', 'Moimenta da Beira', 'Mortágua', 'Nelas', 'Oliveira de Frades', 'Penalva do Castelo', 'Penedono', 'Resende', 'Santa Comba Dão', 'São João da Pesqueira', 'Sátão', 'Sernancelhe', 'Tabuaço', 'Tarouca', 'Vila Nova de Paiva', 'Vouzela'],
    'Viana do Castelo': ['Viana do Castelo', 'Ponte de Lima', 'Arcos de Valdevez', 'Caminha', 'Melgaço', 'Monção', 'Paredes de Coura', 'Ponte da Barca', 'Valença', 'Vila Nova de Cerveira'],
    'Vila Real': ['Vila Real', 'Chaves', 'Peso da Régua', 'Valpaços', 'Alijó', 'Boticas', 'Mesão Frio', 'Mondim de Basto', 'Montalegre', 'Murça', 'Ribeira de Pena', 'Sabrosa', 'Santa Marta de Penaguião', 'Vila Pouca de Aguiar'],
    'Castelo Branco': ['Castelo Branco', 'Covilhã', 'Fundão', 'Sertã', 'Belmonte', 'Idanha-a-Nova', 'Oleiros', 'Penamacor', 'Proença-a-Nova', 'Vila de Rei', 'Vila Velha de Ródão'],
    'Évora': ['Évora', 'Estremoz', 'Montemor-o-Novo', 'Vendas Novas', 'Alandroal', 'Arraiolos', 'Borba', 'Mora', 'Mourão', 'Portel', 'Redondo', 'Reguengos de Monsaraz', 'Viana do Alentejo', 'Vila Viçosa'],
    'Guarda': ['Guarda', 'Seia', 'Gouveia', 'Aguiar da Beira', 'Almeida', 'Celorico da Beira', 'Figueira de Castelo Rodrigo', 'Fornos de Algodres', 'Manteigas', 'Mêda', 'Pinhel', 'Sabugal', 'Trancoso', 'Vila Nova de Foz Côa'],
    'Beja': ['Beja', 'Odemira', 'Serpa', 'Moura', 'Aljustrel', 'Almodôvar', 'Alvito', 'Barrancos', 'Castro Verde', 'Cuba', 'Ferreira do Alentejo', 'Mértola', 'Ourique', 'Vidigueira'],
    'Bragança': ['Bragança', 'Mirandela', 'Macedo de Cavaleiros', 'Alfândega da Fé', 'Carrazeda de Ansiães', 'Freixo de Espada à Cinta', 'Miranda do Douro', 'Mogadouro', 'Torre de Moncorvo', 'Vila Flor', 'Vimioso', 'Vinhais'],
    'Portalegre': ['Portalegre', 'Elvas', 'Ponte de Sor', 'Alter do Chão', 'Arronches', 'Avis', 'Campo Maior', 'Castelo de Vide', 'Crato', 'Fronteira', 'Gavião', 'Marvão', 'Monforte', 'Nisa', 'Sousel'],
    'Ilha da Madeira': ['Funchal', 'Santa Cruz', 'Câmara de Lobos', 'Machico', 'Ribeira Brava', 'Calheta', 'Pontassol', 'Porto Moniz', 'São Vicente', 'Santana', 'Porto Santo'],
    'Ilha de São Miguel': ['Ponta Delgada', 'Ribeira Grande', 'Lagoa', 'Vila Franca do Campo', 'Povoação', 'Nordeste'],
    'Ilhas Açores (Outras)': ['Angra do Heroísmo', 'Praia da Vitória', 'Horta', 'Madalena', 'São Roque do Pico', 'Lajes do Pico', 'Velas', 'Calheta', 'Santa Cruz da Graciosa', 'Vila do Porto', 'Santa Cruz das Flores', 'Lajes das Flores', 'Vila do Corvo']
}

def get_location_info(tags):
    # Try multiple tags for location
    raw_city = tags.get('addr:city') or tags.get('addr:municipality') or tags.get('addr:suburb') or tags.get('addr:province') or ""
    raw_suburb = tags.get('addr:suburb') or tags.get('addr:neighborhood') or ""
    postcode = tags.get('addr:postcode', "")
    name = tags.get('name', "")
    ref = tags.get('ref', "")
    
    # Simple postcode to district mapping (first digit)
    POSTCODE_TO_DISTRICT = {
        '1': 'Lisboa', '2': 'Santarém', '3': 'Coimbra', '4': 'Porto',
        '5': 'Vila Real', '6': 'Castelo Branco', '7': 'Évora', '8': 'Faro', '9': 'Açores/Madeira'
    }
    
    # Additional mapping for common towns/parishes
    TOWN_TO_MUNI = {
        'Vilamoura': 'Loulé', 'Quarteira': 'Loulé', 'Vale de Lobo': 'Loulé', 'Almancil': 'Loulé',
        'Linda-a-Velha': 'Oeiras', 'Porto Salvo': 'Oeiras', 'Alges': 'Oeiras', 'Paco de Arcos': 'Oeiras', 
        'Barcarena': 'Oeiras', 'Carnaxide': 'Oeiras', 'Queijas': 'Oeiras', 'Alfragide': 'Amadora',
        'Ericeira': 'Mafra', 'Malveira': 'Mafra',
        'Fogueteiro': 'Seixal', 'Corroios': 'Seixal', 'Amora': 'Seixal', 'Fernao Ferro': 'Seixal',
        'Pinhal Novo': 'Palmela', 'Quinta do Anjo': 'Palmela', 'Azeitao': 'Setúbal',
        'Estoril': 'Cascais', 'Parede': 'Cascais', 'Carcavelos': 'Cascais', 'Alcabideche': 'Cascais',
        'Alverca': 'Vila Franca de Xira', 'Povoa de Santa Iria': 'Vila Franca de Xira',
        'Leca da Palmeira': 'Matosinhos', 'Senhora da Hora': 'Matosinhos',
        'Sacavem': 'Loures', 'Sacavém': 'Loures', 'Moscavide': 'Loures', 'Bobadela': 'Loures', 'Sta Iria de Azoia': 'Loures',
        'Albufeira': 'Albufeira', 'Fatima': 'Ourém', 'Fátima': 'Ourém',
        'Queluz': 'Sintra', 'Agualva-Cacem': 'Sintra', 'Algueirao': 'Sintra', 'Rio de Mouro': 'Sintra', 'Belas': 'Sintra',
        'Costa de Caparica': 'Almada', 'Charneca de Caparica': 'Almada', 'Laranjeiro': 'Almada', 'Feijo': 'Almada',
        'Parque das Nacoes': 'Lisboa', 'Belem': 'Lisboa', 'Benfica': 'Lisboa', 'Olivais': 'Lisboa',
        'Antas': 'Porto', 'Foz do Douro': 'Porto', 'Boavista': 'Porto',
        'Canidelo': 'Vila Nova de Gaia', 'Mafamude': 'Vila Nova de Gaia'
    }

    
    # 1. Try postcode first (very reliable for district)
    if postcode and len(postcode) >= 1:
        first_digit = postcode[0]
        if first_digit in POSTCODE_TO_DISTRICT:
            district = POSTCODE_TO_DISTRICT[first_digit]
            # Special case for 9 (Madeira/Açores)
            if first_digit == '9':
                if postcode.startswith('95') or postcode.startswith('96'): district = 'Ilha de São Miguel'
                elif postcode.startswith('90') or postcode.startswith('91') or postcode.startswith('92') or postcode.startswith('93') or postcode.startswith('94'): district = 'Ilha da Madeira'
                else: district = 'Ilhas Açores (Outras)'
            
            # Now try to find municipality and parish from other tags
            found_muni = raw_city or district
            found_parish = raw_suburb or raw_city or district
            
            for s in [raw_city, raw_suburb, name, ref]:
                if not s: continue
                # Check TOWN_TO_MUNI
                for town, muni in TOWN_TO_MUNI.items():
                    if town.lower() in s.lower():
                        return district, muni, town
                # Check PORTUGAL_MAPPING
                for d, munis in PORTUGAL_MAPPING.items():
                    for m in munis:
                        if m.lower() in s.lower():
                            found_muni = m
                            if not raw_suburb: found_parish = m
            
            return district, found_muni, found_parish

    # 2. Search strategy: Location tags, Name tag, Ref tag
    search_strings = [raw_city, raw_suburb, name, ref]
    
    for s in search_strings:
        if not s or len(s) < 3: continue
        
        # Check TOWN_TO_MUNI
        for town, muni in TOWN_TO_MUNI.items():
            if town.lower() in s.lower():
                for dist, munis in PORTUGAL_MAPPING.items():
                    if muni in munis:
                        return dist, muni, town
        
        # Check PORTUGAL_MAPPING directly
        s_lower = s.lower()
        for dist, munis in PORTUGAL_MAPPING.items():
            if s_lower == dist.lower() or dist.lower() in s_lower:
                return dist, dist, raw_suburb or raw_city or dist
            for muni in munis:
                if s_lower == muni.lower() or muni.lower() in s_lower:
                    return dist, muni, raw_suburb or raw_city or muni
                    
    return "Outros", raw_city or "Portugal", raw_suburb or raw_city or "Portugal"


def get_clean_brand(tags):
    brand = tags.get('brand') or tags.get('operator') or 'Mobi.E'
    brand_lower = brand.lower()
    
    if 'edp' in brand_lower: return 'EDP'
    if 'galp' in brand_lower: return 'Galp'
    if 'powerdot' in brand_lower: return 'Powerdot'
    if 'ionity' in brand_lower: return 'Ionity'
    if 'tesla' in brand_lower: return 'Tesla'
    if 'continente' in brand_lower: return 'Continente Plug&Charge'
    if 'pingo doce' in brand_lower: return 'Pingo Doce'
    if 'lidl' in brand_lower: return 'Lidl'
    if 'aldi' in brand_lower: return 'Aldi'
    if 'mercadona' in brand_lower: return 'Mercadona'
    if 'ikea' in brand_lower: return 'IKEA'
    if 'auchan' in brand_lower: return 'Auchan'
    if 'kargo' in brand_lower: return 'Kargo'
    if 'prio' in brand_lower: return 'PRIO'
    if 'iberdrola' in brand_lower: return 'Iberdrola'
    
    return brand

inserted = 0

for element in data:
    lat = element.get('lat')
    lng = element.get('lon')
    
    if not lat or not lng:
        continue
        
    tags = element.get('tags', {})

    
    s_id = str(uuid.uuid4())
    name = tags.get('name') or tags.get('ref') or "Posto de Carregamento"
    brand = get_clean_brand(tags)
    
    street = tags.get('addr:street', '')
    housenumber = tags.get('addr:housenumber', '')
    address = f"{street} {housenumber}".strip() or "Portugal"
    
    district, municipality, parish = get_location_info(tags)




    
    # Using lat/lng from the top of the loop

    
    # Posts
    try:
        total_posts = int(tags.get('capacity', 2))
    except ValueError:
        total_posts = 2
        
    available_posts = random.randint(0, total_posts)
    
    # Power
    powers = []
    for k, v in tags.items():
        if k.endswith(':output'):
            powers.append(v)
    
    power_info = ", ".join(list(set(powers))) if powers else ("22 kW" if total_posts <= 2 else "50 kW, 22 kW")
    
    cursor.execute('''
        INSERT INTO stations (id, name, brand, address, district, municipality, parish, lat, lng, type)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
    ''', (s_id, name, brand, address, district, municipality, parish, lat, lng, 'ev'))

    
    cursor.execute('''
        INSERT INTO ev_chargers (station_id, available_posts, total_posts, power_info)
        VALUES (?, ?, ?, ?)
    ''', (s_id, available_posts, total_posts, power_info))
    
    inserted += 1

conn.commit()
conn.close()

print(f"Successfully processed and inserted {inserted} real charging stations from OSM.")

File: process_osm.js
const fs = require('fs');
const path = require('path');
const { getDb, saveDb, uuid, queryAll, queryRun } = require('./db');
const { getLocationInfo, getLocationFromCoords } = require('./location_data');

function getSocketInfo(tags) {
  const sockets = {};
  for (const [k, v] of Object.entries(tags)) {
    if (k.startsWith('socket:')) {
      const parts = k.split(':');
      const socketType = parts[1];
      if (socketType) {
        if (!sockets[socketType]) sockets[socketType] = {};
        if (parts.length > 2) {
          sockets[socketType][parts[2]] = v;
        } else {
          sockets[socketType]['count'] = v;
        }
      }
    }
  }
  return sockets;
}

function getCleanBrand(tags) {
  const raw = tags['brand'] || tags['operator'] || 'Mobi.E';
  const brandLower = raw.toLowerCase();

  const brandMapping = {
    'edp': 'EDP', 'galp': 'Galp', 'powerdot': 'Powerdot', 'ionity': 'Ionity',
    'tesla': 'Tesla', 'continente': 'Continente Plug&Charge',
    'pingo doce': 'Pingo Doce', 'lidl': 'Lidl', 'aldi': 'Aldi',
    'mercadona': 'Mercadona', 'ikea': 'IKEA', 'auchan': 'Auchan',
    'kargo': 'Kargo', 'prio': 'PRIO', 'iberdrola': 'Iberdrola',
    'atlante': 'Atlante', 'repsol': 'Repsol', 'wowplug': 'wowplug',
    'mobiletric': 'Mobiletric', 'helexia': 'Helexia', 'renewing': 'Renewing',
    'segma': 'SEGMA', 'maksu': 'Maksu', 'ecoinside': 'ecoinside',
    'mobismart': 'MobiSmart', 'moon': 'Moon', 'leve': 'leve',
    'moeve': 'Moeve', 'evce': 'EVCE', 'klc': 'KLC',
    'hexagonal ocean': 'Hexagonal Ocean', 'factor energia': 'Factor Energia',
    'cep': 'Cepsa', 'cepsa': 'Cepsa', 'bp pulse': 'Iberdrola',
    'e-flow': 'e-flow', 'porsche': 'Porsche', 'true kare': 'wowplug',
    'charging together': 'Charging Together',
    'loulé concelho global': 'Loulé Concelho Global', 'emel': 'EMEL',
    'ev power': 'EV Power', 'generation journey': 'Generation Journey',
    'dte': 'DTE', 'mota engil': 'Mota Engil',
    'galp geste': 'Galp', 'galp power': 'Galp',
    'prio.e': 'PRIO', 'helexia ii': 'Helexia',
    'maksu services': 'Maksu', 'klc serviços': 'KLC',
  };

  for (const [key, mapped] of Object.entries(brandMapping)) {
    if (brandLower.includes(key)) return mapped;
  }
  return raw;
}

function normalizeStreet(street) {
  const abbreviations = {
    'Av. ': 'Avenida ',
    'Av ': 'Avenida ',
    'R. ': 'Rua ',
    'R ': 'Rua ',
    'Tv. ': 'Travessa ',
    'Tv ': 'Travessa ',
    'Pç. ': 'Praça ',
    'Pc. ': 'Praça ',
    'Pr. ': 'Praça ',
    'Lg. ': 'Largo ',
    'Lgo. ': 'Largo ',
    'Al. ': 'Alameda ',
    'Estr. ': 'Estrada ',
    'Cç. ': 'Calçada ',
    'Calç. ': 'Calçada ',
    'Rot. ': 'Rotunda ',
    'Prct. ': 'Praceta ',
  };
  for (const [abbr, full] of Object.entries(abbreviations)) {
    if (street.startsWith(abbr)) {
      return full + street.substring(abbr.length);
    }
  }
  return street;
}

function parseCapacity(tags) {
  const cap = tags['capacity'] || '';
  const n = parseInt(cap);
  return isNaN(n) ? 2 : Math.max(n, 1);
}

function parsePowerInfo(sockets) {
  const powerParts = new Set();
  for (const [, details] of Object.entries(sockets)) {
    const output = details['output'];
    if (output) powerParts.add(output);
  }
  return powerParts.size > 0 ? [...powerParts].sort().join(', ') : 'N/A';
}

async function main() {
  const osmPath = path.join(__dirname, 'osm_data.json');
  console.log('Loading OSM data...');
  const data = JSON.parse(fs.readFileSync(osmPath, 'utf-8'));
  console.log(`Loaded ${data.length} stations from OSM`);

  const db = await getDb();

  db.run("DELETE FROM ev_chargers");
  db.run("DELETE FROM stations WHERE type = 'ev'");
  db.run("DROP TABLE IF EXISTS ev_chargers");
  db.run(`
    CREATE TABLE ev_chargers (
      station_id TEXT PRIMARY KEY,
      available_posts INTEGER,
      total_posts INTEGER,
      power_info TEXT,
      socket_types TEXT,
      fee TEXT,
      access TEXT,
      opening_hours TEXT,
      phone TEXT,
      website TEXT,
      network TEXT,
      FOREIGN KEY(station_id) REFERENCES stations(id)
    )
  `);

  let inserted = 0;
  let skipped = 0;
  let statsOutros = 0;
  const statsByDistrict = {};

  for (const element of data) {
    let lat = element.lat;
    let lng = element.lon;
    if (!lat || !lng) {
      skipped++;
      continue;
    }

    const tags = element.tags || {};

    let loc = getLocationInfo(tags);
    if (!loc) {
      loc = getLocationFromCoords(lat, lng);
    }

    const [municipality, district] = loc || ['Outros', 'Outros'];
    if (district === 'Outros') statsOutros++;

    statsByDistrict[district] = (statsByDistrict[district] || 0) + 1;

    const cityFromTag = (tags['addr:city'] || '').trim();
    const parish = cityFromTag || municipality;

    const name = tags['name'] || tags['ref'] || `Posto ${municipality}`;

    const brand = getCleanBrand(tags);

    const streetRaw = tags['addr:street'] || '';
    const street = normalizeStreet(streetRaw);
    const housenumber = tags['addr:housenumber'] || '';
    const address = `${street} ${housenumber}`.trim() || `${parish}, ${municipality}`;

    const totalPosts = parseCapacity(tags);
    const availablePosts = totalPosts;

    const sockets = getSocketInfo(tags);
    const socketTypesJson = Object.keys(sockets).length > 0 ? JSON.stringify(sockets) : null;
    const powerInfo = parsePowerInfo(sockets);

    const fee = tags['fee'] || 'yes';
    const access = tags['access'] || 'yes';
    const openingHours = tags['opening_hours'] || '24/7';
    const phone = tags['phone'] || '';
    const website = tags['website'] || tags['contact:website'] || '';
    const network = tags['network'] || 'Mobi.E';

    const sId = uuid();

    db.run(
      `INSERT INTO stations (id, name, brand, address, district, municipality, parish, lat, lng, type)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [sId, name, brand, address, district, municipality, parish, lat, lng, 'ev']
    );
    db.run(
      `INSERT INTO ev_chargers (station_id, available_posts, total_posts, power_info,
                                socket_types, fee, access, opening_hours, phone, website, network)
       VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [sId, availablePosts, totalPosts, powerInfo,
       socketTypesJson, fee, access, openingHours, phone, website, network]
    );

    inserted++;
  }

  saveDb();

  console.log('\n' + '='.repeat(60));
  console.log('RESULTS:');
  console.log(`  Total in OSM: ${data.length}`);
  console.log(`  Inserted: ${inserted}`);
  console.log(`  Skipped (no coords): ${skipped}`);
  console.log(`  Still "Outros" district: ${statsOutros}`);

  const total = data.length;
  const improvement = total > 0 ? (100 * statsOutros / total).toFixed(1) : '0.0';
  console.log(`  Mapping improvement: ${improvement}%`);
}

main().catch(err => {
  console.error('Error:', err);
  process.exit(1);
});
